### DISCUSSION TITLE
Connect with MicroPython

### DISCUSSION THREAD

**Discussion by Felienne** [hedyorg/hedy#4586] [Current student projects]
It would be awesome if we can support some features of [MicroPython](https://micropython.org/) in Hedy, maybe as an extension, especially in connection with the [microbit](https://tech.microbit.org/software/micropython/). Many schools have microbit and combining Hedy's easy and local syntax with a physical tool could be super powerful!
 
It would roughly entail:
* Thinking of what feature to support (through interviews with teachers using micro:bit maybe)?
* Adding some features to the language
* Generating corresponding micro:python
* Making sure a user can download the code (with a new UI element probably)
* Thinking of fun adventures that fit the trajectory

**Comment by Felienne** [comment]
Hi @rmagedon97!

I promised to scope this out a bit more. Let's think of step 1 first: a student types `print hello` in level 1 and can choose to have that program sent to the micro:bit. This involves three steps as far as I can see:

1. Figure out **what** code to send to the micro:bit for printing. I don't know but I guess there is a special Python dialect of library? If you know, let me know here or start a PR where you document it.
2. Figure out **how** to send to the micro:bit. I also don't know but I think it would involve downloading a file and putting that on the micro:bit? For now I think we can suffice by just generating a file, we can deal with the rest later.
3. Change the transpiler such that it generated not only Python dus also microPython or whatever the micro:bit uses. I can help you as soon as I understand what the code should look like but this is wehere the Python is currently generated so that might help?
 
https://github.com/hedyorg/hedy/blob/78e30da12b319ccf52332c51231bf2d9e10cc845/hedy.py#L3560

**Reply by rmagedon97** [reply]
## Questions 
1. To display text on the Micro:Bit LED screen, we can use display.scroll('User_text'). When there are no errors, display.show(Image.HAPPY) can be used, 
    and display.show(Image.ANGRY) can be utilized to indicate an error. There is a library called microbit which we can import using from microbit import *, 
    but it is available exclusively for the MicroPython language. This language is specifically designed for microcontrollers and requires parameters tailored to the specific board, in our case, the Micro:Bit.

2. The Micro:Bit comes pre-flashed with a version of MicroPython. However, we cannot directly run .py scripts on the board. They first need to be converted into a specific .hex
    format. I discovered a GitHub repository that accomplishes this: (https://github.com/GauravBole/microbit-micropython-hex), but I encountered some difficulties using it. There's also the official
    https://github.com/bbcmicrobit/micropython, but it's somewhat complex to operate. I found a simpler 'Python-to-hex' tool (bbcmicrobit/PythonEditor). Specifically,
    the .js script located at PythonEditor/js/python-main.js appears to handle the conversion from .py to .hex efficiently.

3. My idea is that when we click 'Run code', it triggers the normal transpiler. Additionally, there should be another button labeled 'Run on Micro:Bit'. When this button is clicked, the transpiler outputs a
    'microbit' .py script. This script then needs to be converted into a .hex file, which can be donloaded and then directly uploaded onto the Micro:Bit board when connected. If we utilize the tool from 
    https://github.com/bbcmicrobit/PythonEditor, and have this line as a first line 'from microbit import *' in our code, an emulated Micro:Bit board will appears on the screen. We can incorporate this feature into
    our website as an output option for people who don't have the physical board but still want to experiment with it.

**Comment by LegoPink** [comment]
The Microbit website has a python version itself; I think it makes more sense to look at what concepts are needed to jump from Hedy level 18 to the already existing Micro:bit python.

Knowing about classes and imports jump to mind. That would be usefull for PyBricks as well.

**Comment by Felienne** [comment]
> 3\. My idea is that when we click 'Run code', it triggers the normal transpiler. Additionally, there should be another button labeled 'Run on Micro:Bit'. When this button is clicked, the transpiler outputs a
>     'microbit' .py script.

Yes, that is the way to go! Now that we have a tiny demo of code in #5049 we can start to work on the button
